### PR TITLE
Board posts

### PR DESCRIPTION
**Pull request by Mstiekema**
TODO:

- [x] Re-implement Quill
	- [x] Fix bug that text edit headers do not appear multiple times when selecting different posts
	- [x] Make it so the text can be read from area
- [x] BUG: Unable to schedule posts

### CODE REVIEW

**Comment by Mstiekema** [comment]
The `datetime-local` picker for the publish date of the post is not available on every browser [see availability](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local#Browser_compatibility). Maybe look for something else?

**Comment by Riscky** [comment]
> The `datetime-local` picker for the publish date of the post is not available on every browser [see availability](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local#Browser_compatibility). Maybe look for something else?

maybe https://github.com/uxsolutions/bootstrap-datepicker/

**Comment by Mstiekema** [comment]
> maybe [uxsolutions/bootstrap-datepicker](https://github.com/uxsolutions/bootstrap-datepicker/)

This is just a pure datepicker, which is working. I'm looking for a date and time picker in one. If we can't find a proper alternative we can split the date and time tbh.

File: app/views/admin/posts/index.html.haml [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+              .card-body.content
+                = sanitize post.content, tags: %w(strong em u s h1 h2 h3 h4)
+
+
+        = render partial: 'layouts/partials/pagination', locals: {pagy: @pagination}
+
+
+      .col-md-6
+        = form_for @post, :html => { :class => 'form-validation post-content' } do |f|
```
Comment (JustNireon) [inline app/views/admin/posts/index.html.haml:56]: Make this a partial

Comment (JustNireon) [inline app/views/admin/posts/index.html.haml:56]: Make this a partial

File: db/migrate/20200607113340_create_posts.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class CreatePosts < ActiveRecord::Migration[6.0]
```
Comment (JustNireon) [inline db/migrate/20200607113340_create_posts.rb:1]: Title attribute would be nice

File: app/views/members/home/index.html.haml [hunk 3]
```diff
@@ -1,7 +1,47 @@
 - content_for(:page, 'home')
 
 .row
-  .col-lg-6.col-12
+  .col-md-6
+    .card
+      .card-header
+        %i.fas.fa-fw.fa-stream
+        = I18n.t("admin.navigation.messages")
```
Comment (JustNireon) [inline app/views/members/home/index.html.haml:8]: Kind of unnecessary

File: app/assets/javascripts/admin/posts.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+$(document).on("ready page:load turbolinks:load", function () {
+  $(".publish-datetime")[0].flatpickr({
```
Comment (JustNireon) [inline app/assets/javascripts/admin/posts.js:2]: Please make this 24 hr timeclock

File: app/views/members/home/index.html.haml [hunk 5]
```diff
@@ -1,7 +1,47 @@
...
+            %ul
+              %li= I18n.l(post.published_at, format: "%d %B %H:%m")
+              - post.tags&.each do |tag|
+                %li= tag
+          .card-body.content
+            = raw post.content
+    
+    = render partial: 'layouts/partials/pagination', locals: {pagy: @pagination}
+
+  .col-lg-6.col-12.float-right
```
Comment (JustNireon) [inline app/views/members/home/index.html.haml:39]: this seems unnecessary

File: app/views/admin/posts/index.html.haml [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+                  .col-lg-8.col-md-12
+                    = f.label I18n.t("admin.posts.tags")
+                    = f.text_field :tags, :value => @post.tags, :class => 'form-control'
+
+              .form-group
+                .row
+                  .col-lg-4
+                    = f.label I18n.t("admin.posts.published_at")
+                    = f.text_field :published_at, :value => @post.published_at, :class => 'publish-datetime'
```
Comment (JustNireon) [inline app/views/admin/posts/index.html.haml:106]: This does not seem to work

**Review by JustNireon — CHANGES_REQUESTED** [review]
Review for now will continue later

File: app/views/members/home/index.html.haml (hunk 3 above)
Comment (Mstiekema) [reply app/views/members/home/index.html.haml:8]: Idk, I think it looks kinda empty without it.

File: app/views/members/home/index.html.haml (hunk 5 above)
Comment (Mstiekema) [reply app/views/members/home/index.html.haml:39]: The `.col-12` right? Otherwise its required for the info to be displayed on the right side

Comment (JustNireon) [reply app/views/members/home/index.html.haml:39]: Line 39 till line 44 it's double checkout balance now. looks like some old html from the old PR

Comment (Mstiekema) [reply app/views/members/home/index.html.haml:39]: Ah yea, thanks

File: app/views/admin/posts/index.html.haml (hunk 1 above)
Comment (Mstiekema) [reply app/views/admin/posts/index.html.haml:56]: The partial will only be used there then. I can't make a partial and use it on the member side and the admin side. They are different.

**Review by JustNireon — CHANGES_REQUESTED** [review]
LGTM. Only has issue with db:reset / db:setup so this should be fixed or wait for posgres DB

**Comment by Mstiekema** [comment]
Rebase broke:
- [ ] Updating settings (profile img etc.)
- [ ] Quill editor (again :()

**Comment by Mstiekema** [comment]
> LGTM. Only has issue with db:reset / db:setup so this should be fixed or wait for posgres DB

This has been fixed, only need to fix the broken things

**Review by JustNireon — CHANGES_REQUESTED** [review]
Fix navigation icon now it's the same as activities on  the admin backend.
Fix settings spacing on smallers screens it becomes real wonky 
image

**Comment by Riscky** [comment]
As we are changing the branching of Koala (/svsticky/sadserver/issues/323), please change this PR to merge to master instead of development

**Comment by Riscky** [comment]
@Mstiekema any plans on getting this merged? Is there anything blocking you that I can help you out with?

**Comment by Mstiekema** [comment]
> @Mstiekema any plans on getting this merged? Is there anything blocking you that I can help you out with?

I was mostly waiting for postgres and then I ran out of time to work on it. I have to fix the two things Siem mentioned in his comments and check if the migration is still correct with postgres. I will do this sometime this weekend / start of next week. I'll mention you when I'll need another review :)

**Comment by Mstiekema** [comment]
I had some spare time in the train back from work and fixed both things. However now I'm running into Quill not loading (again...), so this will be really be postponed until the weekend.

**Comment by Riscky** [comment]
>  this will be really be postponed until the weekend.

No worries, I'm happy to see progress :)

**Comment by Riscky** [comment]
How's this coming along?

File: app/models/post.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  scope :unpinned, -> { where.not(id: Settings['posts.pinned']) }
+
+  attr_accessor :pinned
+
+  def tags=(tags)
+    write_attribute(:tags, tags.split(' '))
+  end
+
+  def pinned?
+    Settings['posts.pinned'].include? id
```
Comment (Riscky) [inline app/models/post.rb:25]: Should this not just be a boolean in the database instead of storing this in the settings?
This would simplify the code and prevent logic errors

File: app/views/members/home/edit.html.haml [hunk 8]
```diff
@@ -167,6 +167,22 @@
           = I18n.t 'members.home.edit.privacy_panel.view'
           = link_to 'https://public.svsticky.nl/privacystatement.pdf' do
             = I18n.t 'members.home.edit.privacy_panel.privacy_statement'
+      
+      - if !Settings['mailchimp.interests'].blank?
```
Comment (Riscky) [inline app/views/members/home/edit.html.haml:171]: Why does this PR add Mailchimp settings? This should at least belong in a different PR

File: config/app.yml [hunk 9]
```diff
@@ -21,6 +21,9 @@ defaults: &defaults
     committee: ['fotograaf','pr']
     moot: []
 
+  posts:
```
Comment (Riscky) [inline config/app.yml:24]: Again, I don't think this should be a setting, but just a boolean instead

File: config/application.rb [hunk 10]
```diff
@@ -27,7 +27,6 @@ class Application < Rails::Application
     # The default locale is :en and all translations from config/locales/*.rb,yml are auto loaded.
     # config.i18n.load_path += Dir[Rails.root.join('config', 'locales', '**', '*.{rb,yml}')]
     config.i18n.default_locale = :nl
-    config.i18n.fallbacks = true
```
Comment (Riscky) [inline config/application.rb:30]: Why would we want to remove fallbacks? It's better to show a text in the 'wrong' language instead of a translation error in my opinion

File: config/locales/admin.nl.yml [hunk 11]
```diff
@@ -124,6 +125,20 @@ nl:
         pay_to: Kan je dit overmaken naar *NL61INGB0002877106* onder vermelding van _Openstaand %{member}_?
         title: Hey! %{user} van Sticky hier :)
         total_of: Dit geeft een totaal van *%{amount}*
+    posts:
+      author: Auteur
+      content: Bericht tekst
```
Comment (Riscky) [inline config/locales/admin.nl.yml:130]: Just 'tekst' would probably make more sense (or just 'Bericht')

File: config/locales/admin.nl.yml [hunk 12]
```diff
@@ -124,6 +125,20 @@ nl:
...
+      create: Bericht aanmaken
+      edit: Bericht bewerken
+      published_at: Gepubliceerd op
+      status:
+        draft: Concept
+        pinned: Gepinned
+        published: Gepubliceerd
+        review: Review
```
Comment (Riscky) [inline config/locales/admin.nl.yml:138]: What is the 'Review' status for?

File: db/seeds/posts.rb [hunk 13]
```diff
@@ -0,0 +1,32 @@
+require Rails.root.join('db', 'seeds', 'users.rb')
+
+puts '-- Creating posts'
+
+6.times do
+  created_at = Faker::Time.between(from: 2.years.ago, to: Date.yesterday)
+
+  Post.create(
+    title: "Bestuurs bericht",
```
Comment (Riscky) [inline db/seeds/posts.rb:9]: Use  Faker to generate titles and message bodies (and possibly tags)

File: db/seeds/posts.rb [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+      <h5>Dinsdag 13:00: Ode to your CODE (hulpmiddag)</h5>
+      Op 21 januari, een week voor het grote CODE practicum, organiseert Sticky een hulpmiddag in BBG 1.79 van 13:00u tot 17:00u. Hier kan je extra vragen stellen en de laatste hand leggen aan je practicum. Wij zorgen voor koffie, thee en koekjes.
+      <a href="https://koala.svsticky.nl/activities/904">https://koala.svsticky.nl/activities/904</a>
+    EOT
+  )
+
+end
+
+# Set first post to pinned
+Settings['posts.pinned'] = [Post.published&.first&.id]
```
Comment (Riscky) [inline db/seeds/posts.rb:32]: This could crash if no post is published (which is determined randomly)

File: db/migrate/20200607113340_create_posts.rb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+  def change
+    create_table :posts do |t|
+      t.string :title
```
Comment (Riscky) [inline db/migrate/20200607113340_create_posts.rb:4]: I think the title length should have a character limit to prevent display issues (especially on mobile)

File: lib/assets/javascripts/mail.js [hunk 16]
```diff
@@ -0,0 +1,377 @@
+//= require i18n
```
Comment (Riscky) [inline lib/assets/javascripts/mail.js:1]: This seems a bit out of place in this PR?

**Review by Riscky — CHANGES_REQUESTED** [review]
Really nice to see the progress!
It's a huge PR, so I left a lot of comments :)
Most are just questions and small remarks, don't be too intimidated :)

File: app/controllers/members/home_controller.rb [hunk 17]
```diff
@@ -21,20 +21,8 @@ def index
...
-    #      .joins(:participants)
-    #      .where(:participants => { :member => @member }) \
-    #    +
-    #     @member.activities
-    #       .joins(:participants)
-    #       .where("participants.paid = FALSE AND participants.price > 0")
-    #    ).uniq
-    #      .sort_by(&:start_date)
-    #      .reverse!
+    @posts_array = Post.published.pinned + Post.published.unpinned.order(:published_at)
```
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/controllers/members/home_controller.rb:24]: Newest post = top pin

File: app/models/post.rb [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+  enum status: [:draft, :published, :review]
+
+  default_scope { where.not(published_at: nil).order(published_at: :desc) }
+  scope :pinned, -> { where(id: Settings['posts.pinned']) }
+  scope :unpinned, -> { where.not(id: Settings['posts.pinned']) }
+
+  attr_accessor :pinned
+
+  def tags=(tags)
```
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/models/post.rb:20]: Tags can be anything you want it to be. There are no defaults

File: app/views/members/home/edit.html.haml (hunk 8 above)
Comment (Mstiekema) [reply app/views/members/home/edit.html.haml:171]: No clue how this snuck into here. I thought that I had removed all overhead from the original PR

File: config/application.rb (hunk 10 above)
Comment (Mstiekema) [reply config/application.rb:30]: This was from the original PR I think. This line was removed, since the line below it defines the fallbacks. The true here is not correct, the below config is

File: app/javascript/src/posts.js [hunk 19]
```diff
@@ -0,0 +1,15 @@
+// Place all the behaviors and hooks related to the matching controller here.
+// All this logic will automatically be available in application.js.
+import "jquery";
+
+$(document).on("ready page:load turbolinks:load", function () {
+  $(".form-group.post-content").editor({
+    toolbar: [
+      [{ header: [1, 2, 3, 4, false] }],
```
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/javascript/src/posts.js:8]: The `edtior.js` part defines the default quill editor settings. The thought behind this file is if you want to have different toolbars for different parts of koala (e.g. mail, posts, signatures)

File: app/models/post.rb [hunk 20]
```diff
@@ -0,0 +1,43 @@
+# A post is a message viewable by members if published. The author is polymorphic however
+# for now only use from an admin user is possible
+class Post < ApplicationRecord
+  include Sidekiq::Worker
+
+  belongs_to :author, :polymorphic => true
```
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/models/post.rb:6]: Correct

File: app/models/post.rb (hunk 7 above)
Comment (Mstiekema) [reply app/models/post.rb:25]: Fully agree, will fix this soontm

File: config/locales/admin.nl.yml (hunk 12 above)
Comment (Mstiekema) [reply config/locales/admin.nl.yml:138]: If you have created a post but you want another board member to check it, you could set it as "Review". We could also decide to ditch the review state and say that it's the same as being in draft

File: db/seeds/posts.rb (hunk 14 above)
Comment (Mstiekema) [reply db/seeds/posts.rb:32]: Will fix this once pinned posts are a boolean value

File: lib/assets/javascripts/mail.js (hunk 16 above)
Comment (Mstiekema) [reply lib/assets/javascripts/mail.js:1]: Whoopsie, this was left from the merge hell. This file is already present somewhere else

File: app/controllers/members/home_controller.rb (hunk 17 above)
[the remark this answers is not in the document]
Comment (Riscky) [reply app/controllers/members/home_controller.rb:24]: > Newest post = top pin

Do you mean that the newest post is always on top?
Where is that defined?

Also that does not really answer my question: shouldn't we also order the pinned posts?

File: app/javascript/src/posts.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (Riscky) [reply app/javascript/src/posts.js:8]: I suppose that could be useful in the future, but if they're equal now, I don't see why we would need to duplicate the settings.

File: config/locales/admin.nl.yml (hunk 12 above)
Comment (Riscky) [reply config/locales/admin.nl.yml:138]: I don't mind if it stays in, but I doubt people will actually use it

File: app/javascript/src/posts.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/javascript/src/posts.js]: Fixed

File: app/models/admin.rb [hunk 21]
```diff
@@ -14,6 +15,10 @@ def name
     "#{ first_name } #{ infix } #{ last_name }"
   end
 
+  def avatar_representation
+    avatar.representation(resize: '50x50!') if avatar.attached?
```
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/models/admin.rb:17]: It's resized on GET and then cached for 5 min

File: db/migrate/20200607113340_create_posts.rb (hunk 15 above)
Comment (Mstiekema) [reply db/migrate/20200607113340_create_posts.rb:4]: Added a max of 100 chars

File: db/seeds/posts.rb (hunk 14 above)
Comment (Mstiekema) [reply db/seeds/posts.rb]: Fixed

File: app/models/post.rb (hunk 7 above)
Comment (Mstiekema) [reply app/models/post.rb]: Has been fixed

File: app/controllers/members/home_controller.rb (hunk 17 above)
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/controllers/members/home_controller.rb:24]: You can now only have a single post pinned at a time

**Review by Riscky — DISMISSED** [review]
Few nits, but feel free to merge after fixing those

File: app/models/post.rb [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+  def tags=(tags)
+    write_attribute(:tags, tags.split)
+  end
+
+  before_validation(on: :create) do
+    self.published_at = Time.current if published_at.nil?
+  end
+
+  before_save do
+    Post.where.not(id: id).find_each { |post| post.update(pinned: false) } if pinned == true
```
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/models/post.rb:28]: Thanks <3

File: config/app.yml [hunk 23]
```diff
@@ -20,7 +20,7 @@ defaults: &defaults
   additional_positions:
     committee: ['fotograaf','pr']
     moot: []
-
+
```
[the remark this answers is not in the document]
Comment (Mstiekema) [reply config/app.yml]: Fix

File: app/views/layouts/partials/_navigation.html.haml [hunk 24]
```diff
@@ -17,6 +17,10 @@
         %span= I18n.t("navigation.activity")
         %i.fa.fa-ticket-alt{:class => ('active' if controller_name == 'activities')}
 
+      = link_to :posts, :class => 'list-group-item list-group-item-action'do
```
[the remark this answers is not in the document]
Comment (Mstiekema) [reply app/views/layouts/partials/_navigation.html.haml]: Fix
